### PR TITLE
separate stopping ec2 instance from ebs service implementation

### PR DESCRIPTION
**Pull request by xh3b4sd**
This PR is mainly about moving code for preparation of several things. The EBS service implementation in general is tightly coupled with the EC2 Instance management. We do not want this as it makes reconciliation cumbersome. As such we have problems with long running reconciliation loops because of tightly coupled functionality and its implications of error handling. Later we want to have multi master support which would make things even more complicated. Functionality should be resembled and the code starts to be separated more, which is finally what we want.

### CODE REVIEW

File: service/controller/clusterapi/v31/ebs/ebs.go [hunk 1]
```diff
@@ -77,8 +77,11 @@ func (e *EBS) DeleteVolume(ctx context.Context, volumeID string) error {
 	return nil
 }
 
-// DetachVolume detaches an EBS volume. If force is specified data loss may occur. If shutdown is
-// specified the instance will be stopped first.
+// DetachVolume detaches an EBS volume. If force is specified data loss may
+// occur. If shutdown is specified the instance will be stopped first. Note that
+// the long term plan for the interface and functionality is to remove the EC2
+// instance management so the EBS service implementation here only manages EBS
+// volumes.
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v31/ebs/ebs.go]: Just making a note that we want to drop this support from the service implementation at some point.

File: service/controller/clusterapi/v31/ebs/ebs.go [hunk 2]
```diff
@@ -98,7 +101,7 @@ func (e *EBS) DetachVolume(ctx context.Context, volumeID string, attachment Volu
 	}
 
 	if shutdown && wait {
-		e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("waiting for instance %#q to stop", attachment.InstanceID))
+		e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("waiting for instance %#q to be stopped", attachment.InstanceID))
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v31/ebs/ebs.go]: Improving the logs.

File: service/controller/clusterapi/v31/ebs/ebs.go [hunk 3]
```diff
@@ -162,7 +165,7 @@ func (e *EBS) DetachVolume(ctx context.Context, volumeID string, attachment Volu
 		for _, a := range d.Attachments {
 			if a.State != nil {
 				attachmentStatus := *a.State
-				if attachmentStatus != "detached" {
+				if attachmentStatus != ec2.VolumeAttachmentStateDetached {
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v31/ebs/ebs.go]: boyscouting

File: service/controller/clusterapi/v31/resource/tccp/create.go [hunk 4]
```diff
@@ -203,11 +208,9 @@ func (r *Resource) detachVolumes(ctx context.Context, cr v1alpha1.Cluster) error
 			return microerror.Mask(err)
 		}
 
-		// First shutdown the instances and wait for it to be stopped. Then detach
-		// the etcd and docker volume without forcing.
 		force := false
-		shutdown := true
-		wait := true
+		shutdown := false
+		wait := false
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v31/resource/tccp/create.go]: Now that we stop the instance above we do not need to also do it in the EBS service.

File: service/controller/clusterapi/v31/resource/tccp/create_test.go [hunk 5]
```diff
@@ -36,17 +36,12 @@ func Test_Controller_Resource_TCCP_Template_Render(t *testing.T) {
...
-			name: "case 0: basic test",
-			cr:   unittest.DefaultCluster(),
-			ctx:  unittest.DefaultContext(),
-			tp: templateParams{
-				DockerVolumeResourceName:   "rsc-abbacd01",
-				MasterInstanceResourceName: "rsc-ac0dc01",
-			},
+			name:         "case 0: basic test",
+			cr:           unittest.DefaultCluster(),
+			ctx:          unittest.DefaultContext(),
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v31/resource/tccp/create_test.go]: boyscouting

File: service/controller/clusterapi/v31/resource/tccp/resource.go [hunk 6]
```diff
@@ -178,6 +178,66 @@ func (r *Resource) searchMasterInstanceID(ctx context.Context, cr v1alpha1.Clust
...
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("waiting for instance %#q to be stopped", instanceID))
+
+		i := &ec2.DescribeInstancesInput{
+			InstanceIds: []*string{
+				aws.String(instanceID),
+			},
+		}
+
+		err := cc.Client.TenantCluster.AWS.EC2.WaitUntilInstanceStopped(i)
```
Comment (kopiczko) [inline service/controller/clusterapi/v31/resource/tccp/resource.go]: Won't it block operator for too long?

Comment (xh3b4sd) [reply service/controller/clusterapi/v31/resource/tccp/resource.go]: As I tried to clarify in the PR description, all functionality is maintained here. I did not add or remove this. It was already like this. Stopping the instance takes about 30 seconds IIRC. We can think about how to make this better but we have other more important problems actually.

Comment (kopiczko) [reply service/controller/clusterapi/v31/resource/tccp/resource.go]: Ack. Thanks.

**Pull request merged**